1115/healthcheck

henrik-falk opened the pull request.

henrik-falk committed:
add healthz, readyz, livez and resolversz endpoints

henrik-falk committed:
remove altering LDAP mock directory path

henrik-falk committed:
remove exposure through an exception

pep8speaks commented:
Hello @henrik-falk! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`privacyidea/api/healthcheck.py`](https://github.com/privacyidea/privacyidea/blob/27a27e695a5bc386bd212e53a755af1af462fc8d/privacyidea/api/healthcheck.py):

> [Line 60:5](https://github.com/privacyidea/privacyidea/blob/27a27e695a5bc386bd212e53a755af1af462fc8d/privacyidea/api/healthcheck.py#L60): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'

henrik-falk committed:
log exception

henrik-falk committed:
update resolversz endpoint description

henrik-falk committed:
refactor test_resolversz and cover failing resolvers

henrik-falk committed:
fix too long lines

henrik-falk committed:
add hsm readiness check
refactor test for healthz
update livez

henrik-falk committed:
add startupz endpoint
fix PKCS11Mock import error

plettich commented on privacyidea/api/healthcheck.py:
```diff
@@ -0,0 +1,103 @@
+    Health check endpoint that indicates if the app is healthy (running and ready to serve requests).
+    """
+    app = current_app._get_current_object()
+    if not app.config.get('APP_READY'):
```
so currently, when the endpoint is called, the app must already exist. But then the `APP_READY` is always true (unless it set to false again somewhere else).
I guess it is a chicken-egg problem...

henrik-falk replied:
Correct, the APP_READY flag is currently a placeholder until a more sophisticated method of checking if the server has started is developed. In the future, this flag should be switched to False every time dependencies anywhere in the code are not ready.

plettich commented on privacyidea/api/healthcheck.py:
```diff
@@ -0,0 +1,103 @@
+    if not app.config.get('APP_READY'):
+        return send_result({"status": "not healthy"}), 503
+    elif not get_hsm().is_ready:
+        return send_result({"status": "not healthy"}), 503
```
can we return some additional information here, that the app is not healthy due to the HSM? Or is that not intended?

henrik-falk replied:
fixed with commit 680e7655f0874d02c883a1fdcb4dc6647c3104df

henrik-falk committed:
add license info header
add doc string and sphinx documentation
remove unnecessary object referencing
return 503 at exception too
add additional information about readiness
refactor healthz

henrik-falk commented on privacyidea/api/healthcheck.py:
```diff
@@ -0,0 +1,103 @@
+from flask import Blueprint, current_app
```
fixed with commit 680e7655f0874d02c883a1fdcb4dc6647c3104df

henrik-falk commented on privacyidea/api/healthcheck.py:
```diff
@@ -0,0 +1,103 @@
+from flask import Blueprint, current_app
```
fixed with commit 680e7655f0874d02c883a1fdcb4dc6647c3104df

henrik-falk commented on privacyidea/api/healthcheck.py:
```diff
@@ -0,0 +1,103 @@
+    """
+    Health check endpoint that indicates if the app is healthy (running and ready to serve requests).
+    """
+    app = current_app._get_current_object()
```
fixed with commit 680e7655f0874d02c883a1fdcb4dc6647c3104df

henrik-falk commented on privacyidea/api/healthcheck.py:
```diff
@@ -0,0 +1,103 @@
+def check_resolvers(resolvers_list):
+    """
+    Test the connections for the given resolver list.
+    Args:
```
fixed with commit 680e7655f0874d02c883a1fdcb4dc6647c3104df

henrik-falk commented on privacyidea/api/healthcheck.py:
```diff
@@ -0,0 +1,103 @@
+        return send_result(result), 200
+    except Exception as e:
+        log.debug(f"Exception in /resolversz endpoint: {e}")
+        return send_result({"status": "error"}), 500
```
fixed with commit 680e7655f0874d02c883a1fdcb4dc6647c3104df

henrik-falk committed:
get resolver types dynamically from dictionary keys

henrik-falk committed:
refactor resolversz endpoint

henrik-falk committed:
rename according to changes
